core/src/main/kotlin/org/eln2/sim/thermal/main.kt

```diff
@@ -0,0 +1,146 @@
+import kotlinx.coroutines.*
+import kotlinx.coroutines.sync.*
+import kotlin.math.*
+
+class ThermalInterface{
+
+    var mutex = Mutex()
+    var R = 1f //heat resistance at the interface
+    var heatTranfer = 0f //heat transfer during a simulation tick through this interface
+
+    var lastTickComputed = 0 //number of the last tick where the heat transfer was computed. Tick number switch between 0 and 1. This is here to prevent computing twice the heat trasnfered when not needed
+
+
+    fun computeHeatTransfer(dt : Float, tickNumber : Int){
+        var T0 = e0.getT(tickNumber)
+        var T1 = e1.getT(tickNumber)
+        if(tickNumber != lastTickComputed) {
+            heatTranfer = (T1 - T0) / R * dt
+            lastTickComputed = tickNumber
+        }
+    }
+
+    lateinit var e0 : ThermalElement
+    lateinit var e1 : ThermalElement
+}
+
+class SignedThermalInterface(thInterface : ThermalInterface, sgn : Float){
+    var mThermalInterface :ThermalInterface = thInterface
+    var sign : Float = sgn
+
+    fun computeHeatTransfer(dt : Float, tickNumber : Int):Float{
+        mThermalInterface.computeHeatTransfer(dt,tickNumber)
+        return sign*mThermalInterface.heatTranfer;
+    }
+}
+
+class ThermalElement{
+    var C = 1f; //Heat capacity of a given element
+    var P = 0f; //Power dissipated as heat in the thermal element (from electrical circuit)
+
+    //temperature needs to be stored twice: at time $t$, we use T0, then we use that to
+    //compute temperature at time $t+dt$ that we store in T1, then for $t + dt+dt$ we
+    //store it in T0 again etc
+    var T0 = 300f
+    var T1 = 300f
+    //var useT0T1 = 0
+
+    fun getT(tickNumber: Int):Float{
+        if(tickNumber.rem(2) == 0){
+            return T0
+        }
+        else{
+            return T1
+        }
+    }
+
+    suspend fun updateT(dt : Float, tickNumber: Int){
+        //println("c")
+            for (signedInterface in themalInterfacesList) {
+                signedInterface.mThermalInterface.mutex.withLock {
+                    var J = signedInterface.computeHeatTransfer(dt, tickNumber)
+                    if (tickNumber.rem(2) == 0) {
+                        T1 = T1 + J / C + P * dt
+                    } else {
+                        T0 = T0 + J / C + P * dt
+                    }
+                }
+            }
+    }
+
+    var themalInterfacesList = ArrayList<SignedThermalInterface>()
+}
+
+//TODO: change name?
+//helper class that creates thermal elements / interfaces from real world physical values
+class ThermalBuilder{
+
+    fun createWireThermalElement(length: Float, radius : Float):ThermalElement{
+        var wire = ThermalElement()
+
+        var volume = length * PI * radius*radius
+        var volumicMass = 8940f //copper volumic mass in kg/m^3
+        var mass = volume * volumicMass
+        var heatCapacityPerKilogram = 385f //copper heat capacity per Kg in J/Kg/K
```

Comment:
I foresee someone eventually wanting to abstract away these two parameters, but thank you for finding them :)

Reply:
Yup, I'm planning to write some kind of map where you can look for these values for any material we use. For now, I hard coded copper just to show how the API could look like